apps/gdalalg_raster_calc.cpp

```diff
@@ -0,0 +1,617 @@
+ *
+ * Project:  GDAL
+ * Purpose:  "gdal raster calc" subcommand
+ * Author:   Daniel Baston
+ *
+ ******************************************************************************
+ * Copyright (c) 2025, ISciences LLC
+ *
+ * SPDX-License-Identifier: MIT
+ ****************************************************************************/
+
+#include "gdalalg_raster_calc.h"
+
+#include "../frmts/vrt/gdal_vrt.h"
+#include "../frmts/vrt/vrtdataset.h"
+
+#include "gdal_priv.h"
+#include "gdal_utils.h"
+
+#include <algorithm>
+#include <cpl_vsi_virtual.h>
+#include <optional>
+
+//! @cond Doxygen_Suppress
+
+#ifndef _
+#define _(x) (x)
+#endif
+
+struct GDALCalcOptions
+{
+    bool checkSRS{true};
+    bool checkExtent{true};
+};
+
+static bool MatchIsCompleteVariableNameWithNoIndex(const std::string &str,
+                                                   size_t from, size_t to)
+{
+    if (to < str.size())
+    {
+        // If the character after the end of the match is:
+        // * alphanumeric or _ : we've matched only part of a variable name
+        // * [ : we've matched a variable that already has an index
+        // * ( : we've matched a function name
+        if (std::isalnum(str[to]) || str[to] == '_' || str[to] == '[' ||
+            str[to] == '(')
+        {
+            return false;
+        }
+    }
+    if (from > 0)
+    {
+        // If the character before the start of the match is alphanumeric or _,
+        // we've matched only part of a variable name.
+        if (std::isalnum(str[from - 1]) || str[from - 1] == '_')
+        {
+            return false;
+        }
+    }
+
+    return true;
+}
+
+/**
+ *  Add a band subscript to all instances of a specified variable that
+ *  do not already have such a subscript. For example, "X" would be
+ *  replaced with "X[3]" but "X[1]" would be left untouched.
+ */
+static std::string SetBandIndices(const std::string &origExpression,
+                                  const std::string &variable, int band,
+                                  bool &expressionChanged)
+{
+    std::string expression = origExpression;
+    expressionChanged = false;
+
+    std::string::size_type seekPos = 0;
+    auto pos = expression.find(variable, seekPos);
+    while (pos != std::string::npos)
+    {
+        auto end = pos + variable.size();
+
+        if (MatchIsCompleteVariableNameWithNoIndex(expression, pos, end))
+        {
+            // No index specified for variable
+            expression = expression.substr(0, pos + variable.size()) + '[' +
+                         std::to_string(band) + ']' + expression.substr(end);
+            expressionChanged = true;
+        }
+
+        seekPos = end;
+        pos = expression.find(variable, seekPos);
+    }
+
+    return expression;
+}
+
+struct SourceProperties
+{
+    int nBands{0};
+    int nX{0};
+    int nY{0};
+    std::array<double, 6> gt{};
+    OGRSpatialReference *srs{nullptr};
```

Comment:
I assume srs is leaked.  The following (or something along that line) should rather be used.  (we should probably have a ``typdef std::unique_ptr<OGRSpatialReference, OGRSpatialReferenceReleaser> OGRSpatialReferencePtr`` in ogr_spatialref.h)
```suggestion
    std::unique_ptr<OGRSpatialReference, OGRSpatialReferenceReleaser> srs{nullptr};
```

Reply:
or maybe just store it as a non-pointer object ?